src/main/java/org/opentripplanner/routing/graphfinder/PlaceFinderTraverseVisitor.java

```diff
@@ -167,17 +177,49 @@ private boolean stopHasPatternsWithMode(RegularStop stop, Set<TransitMode> modes
       .anyMatch(modes::contains);
   }
 
+  private boolean stopIsIncludedByStopFilter(RegularStop stop) {
+    return filterByStops.isEmpty() || filterByStops.contains(stop.getId());
+  }
+
+  private boolean stopIsIncludedByStationFilter(RegularStop stop) {
+    return (
+      ((filterByStations.isEmpty() || filterByStations.contains(stop.getParentStation().getId())))
+    );
+  }
+
+  private boolean stopIsIncludedByModeFilter(RegularStop stop) {
+    return filterByModes.isEmpty() || stopHasPatternsWithMode(stop, filterByModes);
+  }
+
   private void handleStop(RegularStop stop, double distance) {
-    if (filterByStops != null && !filterByStops.contains(stop.getId())) {
-      return;
-    }
+    /*
+     Do not consider stop if it is not included in the stop or mode filter
+     or if it or its parent station has already been seen.
+     */
     if (
-      includeStops &&
-      !seenStops.contains(stop.getId()) &&
-      (filterByModes == null || stopHasPatternsWithMode(stop, filterByModes))
+      (includeStations && !stop.isPartOfStation() && !stopIsIncludedByStopFilter(stop)) ||
+      (!includeStations && !stopIsIncludedByStopFilter(stop)) ||
+      (stop.isPartOfStation() && !stopIsIncludedByStationFilter(stop)) ||
```

Comment:
This part is a bit hard to follow but I don't know what to do about it.

Reply:
I would put each part in between `||` into a separate method which has a name and possibly a bit of Javadoc.

This doesn't make the logic less complicated but helps someone else understand what each piece does.